test/renderers/input.control.test.tsx

```diff
@@ -0,0 +1,348 @@
+      <InputControl schema={t.context.schema}
+                      uischema={uischema}
+      />
+    </Provider>
+  );
+
+  const control = findRenderedDOMElementWithClass(tree, 'control');
+  t.not(control, undefined);
+  t.is(control.childNodes.length, 3);
+  t.not(findRenderedDOMElementWithClass(tree, 'root_properties_foo'), undefined);
+  /*t.not(findRenderedDOMElementWithClass(tree, 'valid'), undefined);*/
+
+  const label = findRenderedDOMElementWithTag(tree, 'label') as HTMLLabelElement;
+  t.is(label.textContent, '');
+
+  const input = findRenderedDOMElementWithTag(tree, 'input') as HTMLInputElement;
+
+  const validation = findRenderedDOMElementWithClass(tree, 'validation');
+  t.is(validation.tagName, 'DIV');
+  t.is((validation as HTMLDivElement).children.length, 0);
+});
+
+test.skip('hide', t => {
+  const store = initJsonFormsStore(t.context.data, t.context.schema, t.context.uischema);
+  const tree = renderIntoDocument(
+    <Provider store={store}>
+    <InputControl schema={t.context.schema}
+  uischema={t.context.uischema}
+  visible={false}
+    />
+    </Provider>
+);
+  const control = findRenderedDOMElementWithClass(tree, 'control') as HTMLElement;
+  t.true(control.hidden);
+});
+
+test.skip('show by default', t => {
+  const store = initJsonFormsStore(
+    t.context.data,
+    t.context.schema,
+    t.context.uischema
+  );
+  const tree = renderIntoDocument(
+    <Provider store={store}>
+    <InputControl schema={t.context.schema}
+  uischema={t.context.uischema}
+/>
+  </Provider>
+);
+  const control = findRenderedDOMElementWithClass(tree, 'control') as HTMLElement;
+  t.false(control.hidden);
+});
+
+test('single error', t => {
+  const store = initJsonFormsStore(t.context.data, t.context.schema, t.context.uischema);
+  const tree = renderIntoDocument(
+    <Provider store={store}>
+      <InputControl schema={t.context.schema}
+                      uischema={t.context.uischema}
+      />
+    </Provider>
+  );
+  const validation = findRenderedDOMElementWithClass(tree, 'validation');
+  store.dispatch(update('foo', () => 2));
+  store.dispatch(validate());
+  t.is(validation.textContent, 'should be boolean');
+});
+
+test('multiple errors', t => {
+  const store = initJsonFormsStore(t.context.data, t.context.schema, t.context.uischema);
+  const tree = renderIntoDocument(
+    <Provider store={store}>
+    <InputControl schema={t.context.schema}
+                    uischema={t.context.uischema}
+
+    />
+    </Provider>
+  );
+  const validation = findRenderedDOMElementWithClass(tree, 'validation');
+  store.dispatch(update('foo', () => 3));
+  store.dispatch(validate());
+  t.is(
+    validation.textContent,
+    'should be boolean'
+  );
+});
+
+test('empty errors by default', t => {
+  const store = initJsonFormsStore(
+    t.context.data,
+    t.context.schema,
+    t.context.uischema
+  );
+  const tree = renderIntoDocument(
+    <Provider store={store}>
+    <InputControl schema={t.context.schema}
```

Comment:
Indentation messed up (occurs multiple times; actually I don't care that much, but as we do not yet have set-up a formatter, maybe fix this)